controllers/mc_handler.go

```diff
@@ -0,0 +1,190 @@
+		return dummy, err
+	} else {
+		r.Log.Info("MachineConfig already exists")
+		return false, nil
+	}
+}
+
+// Create a new MachineConfig object for the Custom Resource
+func (r *KataConfigOpenShiftReconciler) newMCForCR(machinePool string) (*mcfgv1.MachineConfig, error) {
+	r.Log.Info("Creating MachineConfig for Custom Resource")
+
+	ic := ignTypes.Config{
+		Ignition: ignTypes.Ignition{
+			Version: "3.2.0",
+		},
+	}
+
+	icb, err := json.Marshal(ic)
+	if err != nil {
+		return nil, err
+	}
+
+	if r.FeatureGatesStatus[featuregates.LayeredImageDeployment] {
+		r.Log.Info("Creating MachineConfig for Custom Resource with image")
+		return r.newImageMCForCR(machinePool, icb)
+	} else {
+		r.Log.Info("Creating MachineConfig for Custom Resource with extension")
+		return r.newExtensionMCForCR(machinePool, icb)
+	}
+
+}
+
+// Create a new MachineConfig object for the Custom Resource with the extension
+func (r *KataConfigOpenShiftReconciler) newExtensionMCForCR(machinePool string, icb []byte) (*mcfgv1.MachineConfig, error) {
+	extension := getExtensionName()
+
+	mc := mcfgv1.MachineConfig{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "machineconfiguration.openshift.io/v1",
+			Kind:       "MachineConfig",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: extension_mc_name,
+			Labels: map[string]string{
+				"machineconfiguration.openshift.io/role": machinePool,
+				"app":                                    r.kataConfig.Name,
+			},
+			Namespace: OperatorNamespace,
+		},
+		Spec: mcfgv1.MachineConfigSpec{
+			Extensions: []string{extension},
+			Config: runtime.RawExtension{
+				Raw: icb,
+			},
+		},
+	}
+
+	return &mc, nil
+}
+
+// Create a new MachineConfig object for the Custom Resource with the image
+func (r *KataConfigOpenShiftReconciler) newImageMCForCR(machinePool string, icb []byte) (*mcfgv1.MachineConfig, error) {
+
+	// Get the LayeredImageDeployment feature gate parameters
+	imageParams := r.FeatureGates.GetFeatureGateParams(context.TODO(), featuregates.LayeredImageDeployment)
+	// Ensure at least osImageURL is set
+	if _, ok := imageParams["osImageURL"]; !ok {
+		return nil, fmt.Errorf("osImageURL not set in feature gate %s", featuregates.LayeredImageDeployment)
+	}
+	// If kernelArguments is not set, set it to an empty string
+	if _, ok := imageParams["kernelArguments"]; !ok {
+		imageParams["kernelArguments"] = ""
+	}
+
+	mc := mcfgv1.MachineConfig{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "machineconfiguration.openshift.io/v1",
+			Kind:       "MachineConfig",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: image_mc_name,
+			Labels: map[string]string{
+				"machineconfiguration.openshift.io/role": machinePool,
+				"app":                                    r.kataConfig.Name,
+			},
+			Namespace: OperatorNamespace,
+		},
+		Spec: mcfgv1.MachineConfigSpec{
+			OSImageURL:      imageParams["osImageURL"],
+			KernelArguments: []string{imageParams["kernelArguments"]},
```

Comment:
Ok a CM doesn't seem to allow that... only structured API I can think of is CRD. Would it be possible for each feature to be described with its own CRD ?

Reply:
Right. It's not possible to have this with CM. So for now only the single string containing the entire kernel argument string needs to be provided and it gets used as-is.
CRDs are best imho for structured info.  Once we stablise the features and move it to KataConfig CRD (with versioning of course) we can have a full-fledged struct to handle these better.

Reply:
May be another approach (for now) could be to use a machineconfig yaml and embed it in the configmap. So for the admin the requirement will be to create a machineconfig yaml and then create the configmap out of the machineconfig yaml. We can't allow the admin user to apply the machineconfig directly as the operator needs to control the application and rollback.

Reply:
> Right. It's not possible to have this with CM. So for now only the single string containing the entire kernel argument string needs to be provided and it gets used as-is.

Does this work ? If so, I'm wondering why `KernelArguments` is `string[]` and not just a `string`. In case this works by chance, maybe worth having some `// FIXME` or `// HACK` comment ?

> CRDs are best imho for structured info. Once we stablise the features and move it to KataConfig CRD (with versioning of course) we can have a full-fledged struct to handle these better.

Yeah in the end it will be proper CRD. I was just wondering if it would be help to introduce them right away.

Reply:
> May be another approach (for now) could be to use a machineconfig yaml and embed it in the configmap. So for the admin the requirement will be to create a machineconfig yaml and then create the configmap out of the machineconfig yaml. We can't allow the admin user to apply the machineconfig directly as the operator needs to control the application and rollback.

Yes this would allow the user to set the kernel arguments correctly.

Reply:
> Does this work ? If so, I'm wondering why KernelArguments is string[] and not just a string. In case this works by chance, maybe worth having some // FIXME or // HACK comment ?

Yeah it works.  The configmap value is a string. And currently I take the value and put it as the only entry for string[]. I'll add a //FIXME and indicate the current limitation. 
For example
```
apiVersion: v1
data:
  kernelArguments: "kvm_intel.tdx=1 rhgb quiet"
  osImageURL: quay.io/bpradipt/coco:ocp415
kind: ConfigMap
```
And "kvm_intel.tdx=1 rhgb quiet" is the only element when creating the machineConfig struct

Reply:
Fixed. Now the elements are converted to an array and passed.